### PR TITLE
stm32l4: Fix two race conditions in low power

### PR DESCRIPTION
**Pull request by agkaminski**
- cmp event could occur before reaching wfi causing wakeup event to be missed and sleep taking much longer (until other wakeup event occured),
- arr event could occur between spinlock set and deep sleep causing timer overflow to be missed (causing ~15 s being lost on system timer).

Remove manual SYSTICK pending flag set, not needed anymore as rescheduling is triggered by timer irqs.

JIRA: ISC-199

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (stm32l4).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7m/stm32/l4/timer.c [hunk 1]
```diff
@@ -203,13 +231,13 @@ void _hal_timerInit(u32 interval)
...
 
-	*(timer_common.lptim + lptim_cr) |= 4;
-
 	timer_common.overflowh.f = timer_irqHandler;
 	timer_common.overflowh.n = lptim1_irq;
 	timer_common.overflowh.got = NULL;
 	timer_common.overflowh.data = NULL;
 	hal_interruptsSetHandler(&timer_common.overflowh);
 
+	*(timer_common.lptim + lptim_cr) |= 4;
```
Comment (HBuczynski) [inline hal/armv7m/stm32/l4/timer.c:240]: I assume that this line enables timer. To keep the order of execution, don't we need a memory barrier?

File: hal/armv7m/stm32/l4/timer.c [hunk 2]
```diff
@@ -146,18 +147,43 @@ void timer_setAlarm(time_t us)
...
+		setval = (timerval[1] - 1) & 0xffff;
+	}
+	else {
+		setval = (timerval[1] + (u32)ticks) & 0xffff;
+
+		if (((setval - timerval[1]) & 0xffff) < hal_timerUs2Cyc(1000)) {
+			/* There is too much risk that interrupt will arrive before wfi
+			 * This can happen regardless of previous checks - can be
+			 * caused by a modulo operation few lines above */
+			setval = (timerval[1] + hal_timerUs2Cyc(1000)) & 0xffff;
```
Comment (HBuczynski) [inline hal/armv7m/stm32/l4/timer.c:172]: You can define value from `hal_timerUs2Cyc(1000)` as a const to not invoke function with the same argument two times.

**Review by HBuczynski — COMMENTED** [review]
General, the code LGTM, my comments are rather nitpickings. 
I have not dived into the documentation of this controller so I have only focused on the code structure.

File: hal/armv7m/stm32/l4/timer.c [hunk 3]
```diff
@@ -146,18 +147,43 @@ void timer_setAlarm(time_t us)
 	}
 
 	hal_spinlockSet(&timer_common.sp, &sc);
-	setval = (timer_getCnt() + (u32)ticks) & 0xffff;
+
+	/* Check if there's pending overflow to be handled
+	 * to not lose it should there be next when sleeping */
+	timerval[0] = timer_getCnt();
+	arrpend = (*(timer_common.lptim + lptim_isr) & 2) != 0 ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply hal/armv7m/stm32/l4/timer.c:154]: On read I don't need to

File: hal/armv7m/stm32/l4/timer.c (hunk 1 above)
Comment (agkaminski) [reply hal/armv7m/stm32/l4/timer.c:241]: Not strictly needed here, but won't hurt, added

**Review by HBuczynski — APPROVED** [review]

**Review by anglov — APPROVED** [review]

**Pull request merged**